code/net/ConnectX-5/mlx5.c

```diff
@@ -0,0 +1,509 @@
+
+  if (fw_rev_minor != REG_FW_REV_MINOR_VAL) {
+    TN_DEBUG("Firmware revision minor is %x, expected value is %x", fw_rev_minor, REG_FW_REV_MINOR_VAL);
+    return false;
+  }
+
+  if (fw_rev_major != REG_FW_REV_MAJOR_VAL) {
+    TN_DEBUG("Firmware revision is major %x expected value is %x", fw_rev_major, REG_FW_REV_MAJOR_VAL);
+    return false;
+  }
+
+  if (cmd_int_rev != REG_CMD_INTERFACE_REVISION_VAL) {
+    TN_DEBUG("Firmware revision is %x expected value is %x", cmd_int_rev, REG_CMD_INTERFACE_REVISION_VAL);
+    return false;
+  }
+
+  if (fw_rev_subminor != REG_FW_REV_SUBMINOR_VAL) {
+    TN_DEBUG("Firmware revision is %x expected value is %x", fw_rev_subminor, REG_FW_REV_SUBMINOR_VAL);
+    return false;
+  }
+
+  /**
+   * Step 2.
+   *  Write the physical location of the command queues to the initialization segment.
+   *  If using 32-bit writes, write the most significant word first.
+   *  The nic_interface field is part of the least significant word
+   *  and must be set to zero (Full NIC/HCA driver), as are the log_cmdq_size and log_cmdq_stride fields
+   * */
+  TN_VERBOSE("### Init - step 2");
+
+  void* command_queues_virt_addr;
+  uintptr_t command_queues_phys_addr;
+
+  if (!tn_mem_allocate(CMDQ_SIZE, &command_queues_virt_addr)) {
+    TN_DEBUG("Could not allocate memory for command queues");
+    return false;
+  }
+
+  if (!tn_mem_virt_to_phys(command_queues_virt_addr, &command_queues_phys_addr)) {
+    TN_DEBUG("Could not get a command_queues's physical address");
+    return false;
+  }
+
+  // Since CMDQ_PHY_ADDR_LOW is not 32 bits but 20, the lower 12 bits
+  // of the command_queues_phys_addr should be 0
+  if ((command_queues_phys_addr & BITS(0, 12)) != 0) {
+    TN_DEBUG("The physical address has not the lower 12 bits set to 0!");
+    return false;
+  }
+
+
+  reg_write(device.addr, 0x10, (uint32_t)(command_queues_phys_addr>>32));
+  // shift the command_queues_phys_addr to write only the top 20 bits of the bottom 32 bits
+  reg_write_field(device.addr, 0x14, REG_CMDQ_PHY_ADDR_LOW, ((uint32_t) (command_queues_phys_addr & 0x00000000FFFFFFFF)) >> 12);
+  reg_clear_field(device.addr, 0x14, REG_NIC_INTERFACE);
+  reg_clear_field(device.addr, 0x14, REG_LOG_CMDQ_SIZE);
+  reg_clear_field(device.addr, 0x14, REG_LOG_CMDQ_STRIDE);
+
+  /**
+   * Step 3.
+   * Read the initializing field from the initialization segment.
+   * Repeat until it is cleared (INIT_SEGMENT.initializing become 0).
+   * */
+  TN_VERBOSE("### Init - step 3");
+
+  IF_AFTER_TIMEOUT(INITIALIZING_TIMEOUT * 1000 * 1000,
+         !reg_is_field_cleared(device.addr, REG_INITIALIZING_OFFSET, REG_INITIALIZING)) {
+    TN_DEBUG("INIT_SEGMENT.initializing did not clear, cannot complete init setup");
+    return false;
+  }
+
+  /**
+   * Step 4.
+   * Execute ENABLE_HCA command.
+   * */
+  struct CommandQueueEntry commandQueueEntryEnableHCA;
+
+  memset(&commandQueueEntryEnableHCA, 0, sizeof(commandQueueEntryEnableHCA));  // init all fields with 0
```

Comment:
note that in C you can do `struct CommandQueueEntry commandQueueEntryEnableHCA = {0};` and this will set all fields to 0, without the need for an explicit memset. 😄